### PR TITLE
Fix edge prediction observed-graph protocol

### PR DESCRIPTION
**Pull request by bokveizen**
## Summary

This PR adds an opt-in corrected protocol for `edge_pred` via `--edge_pred_protocol=observed`, while preserving the current DHG-Bench behavior as the default `legacy` protocol.

The observed protocol makes held-out hyperedge prediction cleaner by:

- Building node embeddings from an observed/support hypergraph only, instead of encoding the full original hypergraph.
- Using validation positives from the validation target set rather than reusing training positives.
- Sampling negatives while excluding the full set of true hyperedges, so held-out positives are not accidentally sampled as negatives.
- Recomputing model-specific preprocessing after constructing the support-only graph.

## Motivation

In the current `edge_pred` pipeline, train/validation/test positive hyperedges are split for scoring, but `model.encoding(data)` still receives `data.hyperedge_index` from the full original hypergraph. This means held-out hyperedges can participate in message passing before they are scored.

The current validation positive loader also uses `train_only_pos + ground_train` for `label="pos"`, so validation metrics do not evaluate the intended validation positives. In addition, test negative sampling can miss held-out test positives in its exclusion set.

## Compatibility

The default remains:

```bash
--edge_pred_protocol=legacy
```

Users can opt into the corrected protocol with:

```bash
python main.py --dname=pubmed --task_type=edge_pred --method=EDHNN --is_default=True --edge_pred_protocol=observed
```

Observed split files are written under a separate split-cache directory such as `lib_edge_splits/ind_observed/...`, so existing legacy split caches are not overwritten.

## Validation

Passed:

```bash
python -m unittest discover -s tests -v
python -m py_compile dhgbench/parameter_parser.py dhgbench/lib_dataset/edge_sampler.py dhgbench/lib_dataset/edge_loaders.py dhgbench/lib_utils/exp_agent.py tests/test_edge_pred_observed_protocol.py
git diff --check
```

One-epoch smoke tests on `pubmed + EDHNN` also completed successfully:

- `--edge_pred_protocol=observed`: exit code 0; test AVG AUROC/AP = 0.9923 / 0.9848
- `--edge_pred_protocol=legacy`: exit code 0; test AVG AUROC/AP = 0.6228 / 0.6294

### CODE REVIEW

**Comment by bokveizen** [comment]
Fixes for the edge_pred task, as mentioned in this issue (https://github.com/Coco-Hut/DHG-Bench/issues/5).

**Comment by bokveizen** [comment]
Added a second commit, `ee06f5a Fix edge prediction protocol leakage`, with the final edge_pred protocol cleanup. Summary:

- Deduplicates edge_pred hyperedges by canonical node-set identity before split construction.
- Uses negative sampling that excludes all true positive hyperedges, not just the local context.
- Keeps the observed protocol on support-only message passing and disables observed+ind because that split leaks train positives into the support graph.
- Fixes the edge_pred logit/probability contract: max/maxmin aggregators now return logits, training uses BCE-with-logits directly, and eval applies one sigmoid.
- Makes edge_pred training/eval more robust by evaluating on the final epoch for early stopping and preserving singleton prediction batches with `.view(-1)`.
- Adds focused regression tests for deduplication, split disjointness, negative exclusion, singleton eval, final-epoch early stopping, and the logit contract.

Validation run locally:

```bash
python tests/test_edge_pred_observed_protocol.py
python tests/test_edge_pred_logits.py
python -m py_compile dhgbench/lib_dataset/edge_loaders.py dhgbench/lib_utils/exp_agent.py dhgbench/lib_utils/train_agent.py dhgbench/lib_utils/metrics.py dhgbench/lib_utils/aggregator.py tests/test_edge_pred_observed_protocol.py tests/test_edge_pred_logits.py
```
